src/main/java/org/apache/commons/net/ftp/FTPClient.java

```diff
@@ -1917,6 +1917,7 @@ protected boolean _retrieveFile(final String command, final String remote, final
             Util.copyStream(input, local, getBufferSize(),
                     CopyStreamEvent.UNKNOWN_STREAM_SIZE, mergeListeners(csl),
                     false);
+            input.close();
```

Comment:
The finally block calls `Util.closeQuietly(input);` so why do we need this?

Reply:
It looks like the inputstream needs to be closed before trying to get the transfer response. 
When I stepped through the debugger would hang on FTP.__getReply line 320
`        String line = _controlInput_.readLine();`

Reply:
Please rebase on master for your experiment, I fix possible (if unlikely) socket and input stream leak on socket exception in `org.apache.commons.net.ftp.FTPClient._retrieveFile(String, String, OutputStream)`.

Reply:
I'm still seeing the same behaviour as before. When I retrieve the file against the z/OS FTP server it hangs for 2 minutes and then completes. Against the z/VM FTP server it hangs until the control channel times out after 5 minutes and I get the following stack trace:
` java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.read(BufferedReader.java:182)
	at org.apache.commons.net.io.CRLFLineReader.readLine(CRLFLineReader.java:58)
	at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:320)
	at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:299)
	at org.apache.commons.net.ftp.FTP.getReply(FTP.java:728)
	at org.apache.commons.net.ftp.FTPClient.completePendingCommand(FTPClient.java:1850)
	at org.apache.commons.net.ftp.FTPClient._retrieveFile(FTPClient.java:1920)
	at org.apache.commons.net.ftp.FTPClient.retrieveFile(FTPClient.java:1882)
	at com.ibm.ftp.testing.FTPTimeout.getFile(FTPTimeout.java:123)
	at com.ibm.ftp.testing.FTPTimeout.getFileFTP(FTPTimeout.java:69)
	at com.ibm.ftp.testing.FTPTimeout.main(FTPTimeout.java:41)`

Reply:
Does this happen for all downloads or just some of the time?
What version of z/OS and so on?

Reply:
Happens on all retrieves against z/OS or z/VM (both when requesting multiple files or just one). I get the same behaviour using binary file type  + stream mode transfer as well as EBCDIC file type + block mode transfer.  I do not have access to any distributed hosts which have FTP enabled so cannot test against those, but when I tried against a public FTP server (ftp.mirrorservice.org) I did NOT have the problem. I presume because that FTP server closes the data socket once the transfer is complete (replying with 226) whereas the z/OS and z/VM FTP servers do not (replying with 250). 

The z/OS systems I have tried against are z/OS 2.3 and z/OS 2.4. The z/VM host is at 6.4.0. 

This not an issue when using the FTPS Client or if switching back to commons-net 3.6.